src/transforms/kubernetes_pod_metadata.rs

```diff
@@ -0,0 +1,767 @@
+    sender: Option<mpsc::Sender<(Bytes, Vec<(Atom, Value)>)>>,
+    node_name: String,
+    token: Option<String>,
+    host: String,
+    port: String,
+    client: hyper::Client<HttpsConnector<HttpConnector<Resolver>>>,
+}
+
+impl MetadataClient {
+    fn new(
+        trans_config: &KubePodMetadata,
+        sender: mpsc::Sender<(Bytes, Vec<(Atom, Value)>)>,
+        resolver: Resolver,
+        node_name: String,
+        token: Option<String>,
+        host: String,
+        port: String,
+        tls_settings: TlsSettings,
+    ) -> Result<Self, BuildError> {
+        // Select Pod metadata fields which are extracted and then added to Events.
+        let fields = all_fields()
+            .into_iter()
+            .filter(|(key, _)| {
+                trans_config
+                    .fields
+                    .iter()
+                    .any(|field| field.as_str() == *key)
+            })
+            .map(|(_, fun)| fun)
+            .collect();
+
+        Ok(Self {
+            fields,
+            sender: Some(sender),
+            node_name,
+            token,
+            host,
+            port,
+            client: https_client(resolver, tls_settings).context(HttpError)?,
+        })
+    }
+
+    /// Watches for metadata changes and propagates them to Transform.
+    /// Ok only if Transform/Receiver has been droped.
+    /// Err only if it would always error
+    async fn run(&mut self) -> Result<(), BuildError> {
+        // If watch is initiated with None resource_version, we will receive initial
+        // list of pods as synthetic "Added" events.
+        // https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions
+        let mut version = None;
+
+        // Restarts watch request
+        loop {
+            // We could clear Metadata map at this point, as Kubernets documentation suggests,
+            // but then we would have a time gap during which events wouldn't be enriched
+            // with metadata.
+            match self
+                .watch(version.clone(), self.watch_pods_request(version.clone())?)
+                .await
+            {
+                WatchResult::New(new_version) => version = new_version.or(version),
+                WatchResult::Reload => (),
+                WatchResult::Restart => version = None,
+                WatchResult::Shutdown => return Ok(()),
+            }
+        }
+    }
+
+    /// Watches for pods metadata with given watch request.
+    async fn watch(&mut self, mut version: Option<Version>, request: Request<Body>) -> WatchResult {
+        // Start watching
+        let response = self.client.request(request).compat().await;
+        match response {
+            Ok(response) => {
+                info!(message = "Watching Pod list for changes.");
+                let status = response.status();
+                let mut unused = Vec::new();
+                let mut body = response.into_body();
+                'watch: loop {
```

Comment:
this is a bit hard to read, is there a way we can break this up a bit? Generally, I am not a fan of large socpes with named breaks. This code also kinda drifts to the right. Happy to think through this with you but I'd like to see if we can improve this.